scipy/signal/_whittaker.py

```diff
@@ -0,0 +1,159 @@
+import numpy as np
+from scipy.linalg import solve, toeplitz
+
+
+# TODO:
+# 1) Add parameter `order` for the order of the difference penalty
+# 2) C code for _solve_WH_order2_fast
+# 3) Use sparse/banded arrays for _solve_WH_order_direct
+# 4) GCV for lamb
+# 5) Case weights
+
+def whittaker_handerson(signal, lamb = 0.0):
+    """
+    Whittaker-Handerson (WH) smoothing/graduation of a discrete signal.
+
+    This implements WH of order 2, see [1] and [2]. WH can be seen as a P-Spline
+    (penalized B-Spline) of degree zero for equidistant knots.
+
+    In econometrics, the WH graduation of order 2 is referred to as the Hodrick and
+    Prescott filter (https://doi.org/10.2307/2953682).
+    
+    Parameters
+    ----------
+    signal : ndarray
+        A rank-1 array at least of length 3 representing equidistant data points of a
+        signal, e.g. a time series with constant time lag.
+    
+    lamb : float, optional
+        Smoothing or penalty parameter, default is 0.0.
+
+    Returns
+    -------
+    x : ndarray
+        WH smoothed signal.
+
+    Notes
+    -----
+    For signal of :math:`y = (y_1, y_2, \ldots, y_n)`, WH of order :math:`p=2` with
+    smoothing parameter :math:`\lambda` solves the following optimization problem:
+
+    .. math::
+
+        \operatorname{argmin}_{x_i} \sum_i^n w_i (y_i - x_i)^2
+        + \lambda \sum_i^{n-p} (\Delta^p x_i) \,,
+
+    with :math:`\Delta x_i = x_{i+1} - x_i` and
+    :math:`\Delta^2 x_i = \Delta(\Delta x_i) = x_{i+2} - 2x_{i+1} + x_i`.
+    For every input value :math:`y_i`, it generates a smoothed value :math:`x_i`.
+
+    References
+    ----------
+    .. [1] Eilers, P.H.C. (2003).
+           "A perfect smoother". Analytical Chem. 75, 3631-3636.
+           :doi:`10.1021/AC034173T`
+    .. [2] Weinert, Howard L. (2007).
+           "Efficient computation for Whittaker-Henderson smoothing".
+           Computational Statistics and Data Analysis 52:959-74.
+           :doi:`10.1016/j.csda.2006.11.038`
+    """
+    signal = np.asarray(signal)
+    if signal.ndim != 1:
+        msg = f"Input signal array must be of shape (n,); got {signal.shape}"
+        raise ValueError(msg)
```

Comment:
minor: not hit by tests yet based on local coverage report

Reply:
now covered